rrdd/xcp_rrdd.ml

```diff
@@ -475,31 +475,58 @@ let uuid_blacklist = [
 module IntSet = Set.Make(Int)
 
 let domain_snapshot xc =
-  let uuid_of_domain d =
-    Uuid.to_string (Uuid.uuid_of_int_array (d.Xenctrl.handle)) in
+  let dom_uuid_domid_of_domain dom =
+    let uuid = Uuid.(to_string (uuid_of_int_array (dom.Xenctrl.handle))) in
+    let domid = dom.Xenctrl.domid in
+    let start = String.sub uuid 0 18 in
+
+    if List.mem start uuid_blacklist then
+      None
+    else
+      (* Actively hide migrating VM uuids, these are temporary and xenops
+         writes the original and the final uuid to xenstore *)
+      let path =
+        if Astring.String.is_suffix ~affix:"000000000000" uuid then
+          Some "origin-uuid"
+        else if Astring.String.is_suffix ~affix:"000000000001" uuid then
+          Some "final-uuid"
+        else
+          None
+      in
+      let stable_uuid =
```

Comment:
To avoid deep nesting, maybe make this a function that takes `path` as a parameter and pull it up?